Mathlib/Topology/Algebra/LinearTopology.lean

```diff
@@ -0,0 +1,148 @@
+neighborhoods of 0 is nonempty
+
+## Note on the implementation
+
+The definition of Bourbaki doesn't presuppose, but implies, that a linear topology on a ring `R` is
+a ring topology. However, in some of our lemmas, we already assume that `R` is a topological ring.
+This unnecessary assumption will be made unnecessary by results in the ongoing PR #18437.
+Anyway, the idea will be to first define a topology on `R`, and then
+prove that it makes `R` a topological ring, and that it is a linear topology.
+
+-/
+
+open scoped Topology
+open Filter
+
+namespace IsLinearTopology
+
+section Ring
+
+/-- A topology on a ring is linear if its topology is defined by a family of ideals. -/
+class _root_.IsLinearTopology (R : Type*) [Ring R] [TopologicalSpace R] where
+  hasBasis_twoSidedIdeal : (𝓝 (0 : R)).HasBasis
+    (fun I : TwoSidedIdeal R ↦ (I : Set R) ∈ 𝓝 0) (fun I : TwoSidedIdeal R ↦ (I : Set R))
+
+variable {R : Type*} [Ring R] [TopologicalSpace R]
+
+lemma hasBasis_open_twoSidedIdeal [TopologicalRing R] [IsLinearTopology R] :
+    (𝓝 (0 : R)).HasBasis
+      (fun I : TwoSidedIdeal R ↦ IsOpen (I : Set R)) (fun I : TwoSidedIdeal R ↦ (I : Set R)) :=
+  hasBasis_twoSidedIdeal.congr
+    (fun I ↦ ⟨I.asIdeal.toAddSubgroup.isOpen_of_mem_nhds, fun hI ↦ hI.mem_nhds (zero_mem I)⟩)
+    (fun _ _ ↦ rfl)
+
+theorem hasBasis_ideal [IsLinearTopology R] :
+    (𝓝 0).HasBasis (fun I : Ideal R ↦ (I : Set R) ∈ 𝓝 0) (fun I : Ideal R ↦ (I : Set R)) :=
+  hasBasis_twoSidedIdeal.to_hasBasis
+    (fun I hI ↦ ⟨I.asIdeal, hI, subset_rfl⟩)
+    (fun _ ↦ hasBasis_twoSidedIdeal.mem_iff.mp)
+
+theorem hasBasis_open_ideal [TopologicalRing R] [IsLinearTopology R] :
+    (𝓝 0).HasBasis (fun I : Ideal R ↦ IsOpen (I : Set R)) (fun I : Ideal R ↦ (I : Set R)) :=
+  hasBasis_ideal.congr
+    (fun I ↦ ⟨I.toAddSubgroup.isOpen_of_mem_nhds, fun hI ↦ hI.mem_nhds (zero_mem I)⟩)
+    (fun _ _ ↦ rfl)
+
+lemma mk_of_twoSidedIdeal {ι : Sort*} {p : ι → Prop} {s : ι → TwoSidedIdeal R}
+    (h : (𝓝 0).HasBasis p (fun i ↦ (s i : Set R))) :
+    IsLinearTopology R where
+  hasBasis_twoSidedIdeal := h.to_hasBasis (fun i hi ↦ ⟨s i, h.mem_of_mem hi, subset_rfl⟩)
+    (fun _ ↦ h.mem_iff.mp)
+
+theorem _root_.isLinearTopology_iff_hasBasis_twoSidedIdeal :
+    IsLinearTopology R ↔ (𝓝 0).HasBasis
+      (fun I : TwoSidedIdeal R ↦ (I : Set R) ∈ 𝓝 0) (fun I : TwoSidedIdeal R ↦ (I : Set R)) :=
+  ⟨fun _ ↦ hasBasis_twoSidedIdeal, fun h ↦ .mk_of_twoSidedIdeal h⟩
+
+theorem _root_.isLinearTopology_iff_hasBasis_open_twoSidedIdeal [TopologicalRing R] :
+    IsLinearTopology R ↔ (𝓝 0).HasBasis
+      (fun I : TwoSidedIdeal R ↦ IsOpen (I : Set R)) (fun I : TwoSidedIdeal R ↦ (I : Set R)) :=
+  ⟨fun _ ↦ hasBasis_open_twoSidedIdeal, fun h ↦ .mk_of_twoSidedIdeal h⟩
+
+instance [IsLinearTopology R] : Nonempty { J : TwoSidedIdeal R | (J : Set R) ∈ 𝓝 0} := by
+  obtain ⟨J, hJ, _⟩ :=
+    ((hasBasis_twoSidedIdeal (R := R)).mem_iff' Set.univ).mp (Filter.univ_mem)
+  exact ⟨J, hJ⟩
```

Comment:
This does not require `IsLinearTopology`: the whole ring is always a two-sided ideal belonging to any filter. Do you really need this for some reason? If that's in a follow up PR, could you remove it from this one and we'll talk about it later?

Reply:
For whatever lemma in #15007, Lean wanted that instance. I can remove it for the moment.